### ISSUE TITLE
Default Unhandled Rejection Behavior

### ISSUE THREAD

**Issue by benjamingr** [nodejs/promises#26]
Continuation of https://github.com/nodejs/node/issues/830

After a few months of `process.on("unhandledRejection"` event I'm convinced we need a saner default.

Just to be clear - this is about the default behaviour of a rejected promise with no `catch` handler (or `then` handler with a function second argument) attached. Basically - anything that fires

**The different approaches**
-------
 - **Throwing an error for `process.on("uncaughtException"` to deal with.** - this seems impractical now since the hook has been live in the ecosystem for a while. A flag might solve this.
 - **Causing the process to exit.** - this seems impractical now since the hook has been live in the ecosystem for a while. A flag might solve this.
 - **Do nothing by default.** - what we currently do - a lot of complaints about this.
 - **Logging the rejection to the console.** - this is what most browsers do now (namely Chrome) and what bluebird and userland libraries do. I think we should explore that path.

## Proposal

When an `unhandledRejection` event is fired if no handlers are attached the error is logged to stderr. The current behavior is confusing to users and makes it appear as if promises swallow errors. While the `unhandledRejection` hook has been used a lot in the wild a lot of users (even experienced node users) can forget it and get surprising behavior.

People will still be able to opt-out of it by installing any `"unhandledRejection"` handler. 

This aligns Node's behavior with browsers and other environments as well as userland solutions.

**Comment by petkaantonov** [comment]
This is reasonable, the current behavior is really confusing - especially to people who are either new to promises or are callback users that are interacting with a promise returning library

**Comment by Fishrock123** [comment]
Ugh do I really need to make my argument here rather than a PR with actual code? Fine, give me a couple days or something.

**Comment by spion** [comment]
Baring the post-mortem issue (which remains unfixed) and potential shared state cleanup issues, I think this is indeed the biggest problem users have with promises. They start using them, and when the first thrown error occurs there is absolutely no feedback whatsoever as to what happened.

Browsers can add all sorts of fancy mechanisms to log (and unlog) errors and keep track of currently pending promises, unhandled rejections and so on. Unfortunately by default node only has stdout and stderr.

So yes, I think the most sensible thing to do by default is to at least log a warning, but let expert users add their custom unhandledRejection handler to tweak this behaviour to their liking.

Theres been a split in the promise community on this issue, with people advocating any of

* interactively log and un-log unhandled rejections in some sort of UI
* add `.done` and teach `.done` first
* log on the next tick.

Since we don't have UI and don't have built in `.done`...

**Comment by benjamingr** [comment]
@Fishrock123 

> Ugh do I really need to make my argument here rather than a PR with actual code? Fine, give me a couple days or something.

If you'd like to pursue a PR for adding the default handler that logs by all means go for it and I'll gladly help review it.

**Comment by Fishrock123** [comment]
Ok never mind, I guess I feel strongly enough about it to write it now.
​
>  - **Do nothing by default.** - what we currently do - a lot of complaints about this.
​
Agreed, this isn't even a solution; we should do *something*.
​
>  1. **Throwing an error for `process.on("uncaughtException"` to deal with.** - this seems impractical now since the hook has been live in the ecosystem for a while. A flag might solve this.
 2. **Causing the process to exit.** - this seems impractical now since the hook has been live in the ecosystem for a while. A flag might solve this.
 3. **Logging the rejection to the console.** - this is what most browsers do now (namely Chrome) and what bluebird and userland libraries do. I think we should explore that path.
​
These 3 seem to conflate two things: a user-hookable event (`process.on('unhandledRejection')`) and the default action if there are no listeners.
​
[1] is impractical because that means changing the event API, and I'm not sure it even is a default action, but if an error within that handler will cause a process exit, it ends up being the same as [2], except that it is also a breaking change.
​
[2] isn't impractical since it's not actually a change to the API, which would be more difficult because people are most likely using it already. (We can just do *something* if there is no listener, that's not necessarily a breaking change.)
​
[3] isn't very node-like. We're not a browser, and we've long decided to exit on all other unhandled errors.
​
---
​
As a potential (and even past) promises *in node* user, I'd like uncaught/unhandled errors within promises to be **consistent** with other errors.
​
Again, we've long decided to exit on unhanded errors. I'm not exactly sure since when (other than that it was very early), or what the original reasoning was (though I can make guesses). Either way it is quite deeply baked into the node ecosystem and is (from my personal experience) probably what most *node* developers expect. After all, you are more likely to expect something to happen everywhere if something happens at a platform level somewhere. 
​
Why should this be different, just because it's a different API? 
It's not like domains, because domains are like registering a handler for the event. Domains are not exactly a default action.
​
Sure, maybe we are "wrong" about unhandled errors in promises from a browser perspective, but we are "wrong" about all unhandled errors from a browser perspective ([see this twitter thread](https://twitter.com/domenic/status/695339581613297665)). As such, It's not very compelling to me why we would be different here. We've pretty clearly seen unhandled JavaScript errors as something the host platform should decide for a long time.
​
I would prefer it consistent. To me there is no point in trying to continue running if I wrote the program wrong, and compared to a lot of platforms Node doesn't take much to restart. Some basic process managing can probably handle this in production, and has the upside of it being easier to detect that something is wrong rather than searching through error logs. Process managing is already a node paradigm whether you think it's ridiculous or not, and you're still going to need it in case of synchronous errors.
​
Now, I don't think we should exit the process in the current `unhandledRejection` path, but rather add a hook to "throw" on GC, if there is no default handler. You can see my current work for it at https://github.com/nodejs/node/pull/5292 -- but I need to fix it up so that it actually works like that. I will be making a new PR.

---
​
The only way I can see just warning on promise errors is if we do the same from all other JavaScript errors. That is pretty impractical though, and the node developer in me doesn't even like it. Plus, it would be harder to detect potential bugs in production as I outlined above. You cannot assume people write perfect programs. :)

**Comment by Fishrock123** [comment]
There's also a thing here about explicitness -- many of do not find ignoring errors a light thing. We wanted to have a flag to have unhandled errors (any kind) only stop the call stack and warn, without crashing the process, that may be viable, if it is really what you wish for.

**Comment by petkaantonov** [comment]
Logging is better than not doing anything and backwards compatible

**Comment by benjamingr** [comment]
@Fishrock123 thanks. 

Some clarifications: As far as I'm aware the decision to abort on all errors was partially made because errors leave node in an indeterminate state. This is not the case with promise errors generally - if a promise rejects the process generally keeps functioning correctly. This is not the case with unhandled rejections though - those imply a promise chain has been broken and no one was there to recover from an error and those can definitely leave the process in indeterminate state. 

The problem is `unhandledRejection` can in edge cases false-positive which means you abort the process for something that gets handled later.

Code can be written in a way that does not cause an intermediate `unhandledRejection` and a `rejectionHandled` and the ecosystem mostly does so - but it can be risky to impose this on everyone and is opinionated. It's a decision the node CTC _does_ get to make but it's one it should not make lightly. __I__ am perfectly fine with aborting on unhandled rejection if no handler is present - it would certainly help as a first step to bridge the gap with the post-mortem folk - but it's certainly something people object to and it could certainly break some programs in node.  

Also - it's not like logging does not have a precedent - we have a whole warning infrastructure for things that are deprecated, or are possible programming mistakes (like event emitters having too many handlers). Between logging and doing nothing - I'd go for logging. 

As for GC - it suffers from false negatives the same way the current approach suffers from false positives. I think relying on observable GC to perform something drastic like abort the process is extremely slippery slope. 

---------------------

I'd like to (Try to) summon some people, should they wish to participate @erights , @domenic , @kriskowal  to represent the counter-stance.

**Comment by ljharb** [comment]
There is nothing wrong with having a rejected promise with no handler, that 10 minutes later receives a rejection handler - that's why browsers have hooks to "un-warn" about unhandled rejections. Exiting the process by default is a non-starter because that conflicts with the actual language semantics and expected uses.

**Comment by benjamingr** [comment]
@ljharb well, yes _theoretically_ there is nothing wrong with it if it does get handled 10 minutes later. The problem is that it's impossible to determine _ever_ when the error gets handled - it's easily reducable to the halting problem

It's entirely within the power of the node CTC to decide that in node programs such promise usage (resolving after 10 seconds) would cause the process to abort by default (with an opt-out). It's definitely an opinionated take (which is why I pinged domenic, mark and kris) who IIRC object to it.

I'm +1 on logging by the way since it's easy and has an easy opt out. 

Saying something conflicts with actual language semantics and deciding what expected usage is for Node will likely not yield productive discussion - some members of the node CTC feel ignored by TC39 anyway - so I suggest we focus on problems (promises resolving after 10m) rather than telling Node what it can or cannot do.

**Comment by Fishrock123** [comment]
> Some clarifications: As far as I'm aware the decision to abort on all errors was partially made because errors leave node in an indeterminate state. This is not the case with promise errors generally - if a promise rejects the process generally keeps functioning correctly. This is not the case with unhandled rejections though - those imply a promise chain has been broken and no one was there to recover from an error and those can definitely leave the process in indeterminate state. 


Not show how worthwhile this is of discussion, it seems to depend on one's definition of "indeterminate state". I think it is probably good to err on the side of safety, people may not build robust programs.

> The problem is `unhandledRejection` can in edge cases false-positive which means you abort the process for something that gets handled later. (...)


 (Also re: @ljharb)
This is why we should not exit in the existing hook, but rather on GC. If GC happens and it is still unhandled there is no hope of handing it. Unfortunately, this would be easier with reference counting but JS vm's avoid that for perf reasons afaik. See below for more info.

> Also - it's not like logging does not have a precedent - we have a whole warning infrastructure for things that are deprecated, or are possible programming mistakes (like event emitters having too many handlers). Between logging and doing nothing - I'd go for logging. 


In theory; logging to warn is currently very inconsistent, but hopefully that will improve with Jame's warnings PR.

> As for GC - it suffers from false negatives the same way the current approach suffers from false positives. I think relying on observable GC to perform something drastic like abort the process is extremely slippery slope. 


Could you define "false-negative" here?

Let me clarify my plan (which I think I can get working..):

1. If a promise is GC'd and still unhandled, log the stack and exit the process.
2. If we are exiting normally, check that there are no unhandled promise rejections, since GC may not have cleaned them yet. Follow the suit of [1] if there are.

**Comment by benjamingr** [comment]
> Could you define "false-negative" here?

```js
var settings = file.readAsync("nonexisting.json").then(JSON.parse); // let's say the file doesn't exist
// rest of the code here
```

Since this is in global module scope - it would probably never be garbage collected.

**Comment by Fishrock123** [comment]
@benjamingr right, I experienced it and though about it, which is why I'm going to do my best to ensure my [2] is possible. :)

**Comment by ljharb** [comment]
@benjamingr fwiw, in that case i was speaking as a JS developer - I would be very displeased if my node process exited because I intentionally deferred handling of an unhandled rejection.

@Fishrock123 totally agree that if it exits on GC there's no conflict with my expectations, and it'd be a good time to exit by default.

**Comment by chrisdickinson** [comment]
Catching up now — my opinion has pretty much come full circle since the start of this conversation. The TL;DR is that I think that @Fishrock123 is on the right track by making Node crash if an unhandled rejection is garbage collected.

To expand on this: it's tempting to say we should take some default action on unhandled rejection and that programs can be universally written to avoid hitting the unhandled rejection handler. It's especially tempting because it looks like it gives @nodejs/post-mortem an out — in that "promise rejects, is unhandled, we crash", at first glance, looks like it lines up with what they need. However, we always need to wait until next tick to fire the unhandled rejection handler, because doing otherwise causes promises to exhibit confusing behavior from users' perspective — we _always_ want to give synchronous rejections a chance to be handled on the same tick before firing unhandled rejection. This means that we _always_ have to wait until next tick, so crashing on unhandled rejection does not help post-mortem users. 

Crashing the program on unhandledRejection in the absence of a user-installed listener, as mentioned by others, introduces an artificial difference between Node Promises and browser Promises — it invalidates patterns of use that are valid in browser. This is not great for promise users, and should be avoided if we can help it.

Logging is a weird one — it _seems_ like the most innocuous, but consider that many applications use JSON-based logging, like [`bole`](http://npm.im/bole). Node logging through a separate channel means that Node will produce non-JSON output, no matter the desire of the app developer, unless they install an (otherwise meaningless) listener for the `unhandledRejection` event. This problem also applies to `process.emit('warning')`.

My recommendation would be to crash on unhandled rejection GC, per @Fishrock123's plan. I'm not sure about crashing the program on exit if there are unhandled rejections, since it may be that folks are eagerly replacing a rejected promise — we'd have to investigate that. WHATWG streams _may_ do this, IIRC.

With regards to @benjamingr's false negative example — if that were run at global scope it would never crash the program. Running it at module scope, however, would crash at some point after the module has been run. You _could_ get it to exhibit false-negative behavior from a module by exporting the promise, requiring the module, and never handling the promise, but that seems like it would be a fairly uncommon case.

**Comment by petkaantonov** [comment]
Node (and v8) already writes to stderr for plenty of reasons regardless of whether you are using JSON-based logging or not. And some of those cases are not even configurable. Having something that you *can* configure only *default* to stderr must therefore not be as a big problem as you make it out to be.

I am not sure if everyone realizes how the GC works but it is not reliable and it doesn't guarantee that an object is ever collected. I don't consider this acceptable unless you  also add something that reports all unhandled rejections on process exit so that there is some guarantee.

**Comment by Fishrock123** [comment]
> I am not sure if everyone realizes how the GC works but it is not reliable and it doesn't guarantee that an object is ever collected. I don't consider this acceptable unless you also add something that reports all unhandled rejections on process exit so that there is some guarantee.

I quite agree.

> I'm not sure about crashing the program on exit if there are unhandled rejections, since it may be that folks are eagerly replacing a rejected promise — we'd have to investigate that. WHATWG streams may do this, IIRC.

If there is something that would cause this to report "false positives", I'm not really sure. I'm beginning to think that promises necessitate JS vm's to do reference counting so that we can properly detect when something goes out of scope. :(

**Comment by petkaantonov** [comment]
If a program exits with an unhandled rejection that would have been handled had that program exited only a bit later, is it really a handled rejection?  I think it depends on arbitrary philosophy. I personally just avoid the headaches of that by a simple work around:

```js
promise.catch(noop);

setTimeout(() => {
    promise.then(function() {

    });
}, 5000);
```

If `promise` is rejected before the timeout, nothing happens. When the timeout executes, the `then` is a separate branch and it will therefore behave exactly as if that first `catch` had never existed and you will get unhandled rejection. So it's not like use cases like this become impossible.

**Comment by Fishrock123** [comment]
Update: so far I'm unable to get WeakCallbacks on promises to fire in a situation that doesn't use `--expose-gc`. Unsure why, but I sure wish we had reference counting.

**Comment by trevnorris** [comment]
Here's a test case I wrote up as a native module to demonstrate that rejected handles don't trigger the weak callback: https://github.com/nodejs/node/pull/5292#issuecomment-189526572

**Comment by chrishiestand** [comment]
As a node user (not a contributor) who's been using async/await (via babel) for about 6 months, I am a big proponent of node crashing on unhandled promise rejection but I have no strong opinions on the mechanism. This would match the current behavior of an uncaught error, which feels like the semantic equivalent of not catching a rejected promise. Just like any other error, a rejected promise would need a `try/catch` or a `.catch()` to prevent a node crash. And this reinforces a current (best?) practice: catch when you don't want your program to crash - do not catch if you prefer your program bomb out with a stack trace (or if you want the _calling_ functions to implement any catches).

I think backwards compatibility should not be an important factor here - major version releases are for (useful) breaking changes. This is a very worthwhile breaking change, imo. Plenty of warning via communication and documentation will help prepare users.

**Comment by williamkapke** [comment]
Please,please,please do not change the existing behavior of the current native Promises. Letting rejections be swallowed is a feature- not a bug. I have a TON of production code that leverages this intentionally. Think: Fire & Forget code... there is a genuine purpose for it.

*YES*- I have missed a `catch` where there should have been one.
*YES*- It has frustrated me.
... but it isn't Node's fault and it isn't Promise's fault. It is MY fault! At best, I guess it would be nice if I could turn on a flag to have them logged to help me find where I missed it. But I wouldn't want that logging spit out in the middle of my bunyan logs in production.

> Do nothing by default. - what we currently do - a lot of complaints about this. ​

IMHO, These complaints seem equivalent to the complaints about `this` not being the `this` that was expected.

**Comment by aoberoi** [comment]
Is a solution that is exactly like the browsers (specifically: make hooks available for warn and un-warn, and when those hooks aren't used, warn and un-warn by writing to stderr) not viable for some reason?

i think exiting the process would be a little too extreme given @ljharb's [comment](https://github.com/nodejs/promises/issues/26#issuecomment-188879863) about intended semantics of Promise. there are situations where not handling is exactly what a programmer may have intended because she knows how she wants to handle it later. that technique should not be "taxed" by having to structure the handlers in some specific way in node.

if you want control of the stderr output (such as needing it to be JSON formatted), attach the optional handlers and do what you want with the error. i would rather pay that "tax" than losing on explicit warnings as my default, or forcing the techniques such as above to be ruled out and superficially make my JavaScript not portable to node/browser runtimes.

**Comment by Fishrock123** [comment]
If you known a way of clearly letting users know that they have an error that won't impossibly get lost in logs that works cross platform, other than exiting, please let me know. I don't know of such a thing.

**Comment by jrop** [comment]
@Fishrock123 I haven't tested this, but is this what you're looking for?: [loud-rejection](https://github.com/sindresorhus/loud-rejection).

**Comment by yosuke-furukawa** [comment]
I would like to progress this discussion. Current behavior is really confusing for Node and Promise users.

My first impression, I just add flags option like `--throw-error-unhandled-rejection` / `--trace-error-unhandled-rejection`, and if `--throw-error-unhandled-rejection` is on, node throws fatal error, if `--trace-error-unhandled-rejection` is on, node outputs the error to stderr.

Here is the example(currently i have not written `--trace-error-unhandled-rejection` option yet).

https://github.com/nodejs/node/compare/master...yosuke-furukawa:feature/promise/throw-error-unhandled-rejection?expand=1

I don't read the full discussion in this thread. So my opinion may be irrelevant, but I would like to settle down this discussion until v6 released. 

IMHO, if code throws `unhandledRejection`, the code might have a bug. I would like to check the `bug` as soon as possible.
I agree with `we should not change default behavior`. But under the flags, we can change the behavior.

**Comment by benjamingr** [comment]
@yosuke-furukawa you can change the default behavior by doing:

```js
process.on("unhandledRejection", (e) => {
    // handle error
});
```

Although personally I'm strongly against not having a saner default like logging.

**Comment by yosuke-furukawa** [comment]
@benjamingr 

Yes. But I don't want to add the `unhandledRejection` listeners on `every our products and tests`. I need a saner behavior in Node core.

> Although personally I'm strongly against not having a saner default like logging.
Agree, but saner default definition is quite difficult. everyone has different context. so I would like to add flags and not change default behavior.

**Comment by benjamingr** [comment]
> Agree, but saner default definition is quite difficult. everyone has different context. so I would like to add flags and not change default behavior.

Our default behavior deviates from all browsers and other languages at this point and was only decided because it was the easiest course of action and we really wanted to get unhandled rejection tracking merged already.

**Comment by yosuke-furukawa** [comment]
@benjamingr 

I don't have a strong opinion which default behavior is better like `throw error on GC` or `logging comment`. (So this thread is not suitable for my discussion...)

I would like to change the default behavior under my option (`--throw-unhandled-rejection`/`--trace-unhandled-rejection`). these options are similar to `--throw-deprecation` / `--trace-deprecation` / `--trace-sync-io` / `--abort-on-uncaught-exception`.

If the option is on, we can detect our promise error sooner.
If the option will be same to the next default behavior, we just drop the option on next major release.

Anyway, I would like to detect something wrong sooner and I do not want to add the codes like [this](https://github.com/yosuke-furukawa/eater/blob/master/lib/abortPromiseException.js) .

If this change is reasonable, I would like to send a PR to node.

https://github.com/nodejs/node/compare/master...yosuke-furukawa:feature/promise/throw-error-unhandled-rejection?expand=1

**Comment by Fishrock123** [comment]
I have updated my proof-of-concept PR: https://github.com/nodejs/node/pull/5292

It now properly throws on GC of an unhandled promise, and also exits with 1 and prints a strack trace, if there is still an unhandled rejection when the process ends. I'm not yet sure how to have it properly track more than one for that purpose.

**Comment by benjamingr** [comment]
@Fishrock123 ok, as long as we acknowledge and document that this approach has false negatives I think it's definitely a big improvement over the status quo. 

I'm also in favor of warning on unhandled promises that have been unhnandled for a while (even an arbitrary "10 seconds" is better than nothing). 

In addition, node should log all currently unhandled rejections when the process exists (although there might be false positives there).

**Comment by ljharb** [comment]
@benjamingr browsers can retroactively remove console logs for later-handled promises - node can't. How would this be handled?

**Comment by benjamingr** [comment]
@ljharb I've literally never had that impact me and a console log gets removed - but Node can just emit two logs ("possibly unhandled" and "handled") although the easiest thing would be to log only on promises that have not been handled for a while (not a microtick, but rather 2-3 seconds).

**Comment by ljharb** [comment]
Would there be some kind of identifier, so that log processing tools can link the "possibly unhandled" and the "later handled" log entries?

The problem is that it's perfectly valid to add a handler for a promise even an hour, or a day later. Edge case though it may be, anything that purports to deal with unhandled rejected promises needs to account for this.

**Comment by benjamingr** [comment]
> Would there be some kind of identifier, so that log processing tools can link the "possibly unhandled" and the "later handled" log entries?

We already have that - the reference to the promise. We can generate a reference based on that pretty easily. 

> The problem is that it's perfectly valid to add a handler for a promise even an hour, or a day later. Edge case though it may be, anything that purports to deal with unhandled rejected promises needs to account for this.

Why? I think Node can take an opinionated stand and log these - just like Chrome takes an opinionated stand and logs them. It's incredibly useful and when platforms don't do it it's very complicated to use promises reasonably.

It's just the _default_ behavior which can change. Browsers and engines prefer practical cases to theoretical ones all the time as long as whatever behavior you want can be accomplished.

Heck, this is ages away from what most platforms do, in .NET if you have a _pending_ asynchronous request in the context of a server request it aborts with an error by default.

**Comment by jrop** [comment]
@benjamingr I actually agree with @ljharb on this one.  While the "opinionated" way of reporting promises can absolutely be useful, it could be misleading in that it is completely possible to report "unhandled" rejections that actually get handled later on.  I think the previous suggestions of checking the promise's status on GC cleanup, because that would provide consistent "correct" behavior.  However I know nothing of how v8 does GC.

**Comment by Fishrock123** [comment]
I have made a finalized PR of my idea, working in the cases I could think of, possibly ready for v6. Please review with a note of my above statements in this thread.

https://github.com/nodejs/node/pull/6375

**Comment by ljharb** [comment]
@benjamingr Chrome does not take an opinionated stance, in that if 10 days later I add a handler, the log will still be retroactively removed from the console. The same can't be done in node's console output piped elsewhere. Similarly, the reference to the promise can't be used once the console output is piped somewhere.

**Comment by benjamingr** [comment]
@ljharb sure it does - because it can' tell what is actually an unhandled rejection - if you change tabs for example you've seen an unfaithful unhandled rejection although it "would have been" handled if you didn't manually interrupt it.

Similarly in Node, when you exit the program if you get a list of unhandled rejections - there might be some false positives there that actually haven't been errors.

**Comment by benjamingr** [comment]
Hey, just pinging everyone here about two (complementary IMO) proposals: https://github.com/nodejs/node/pull/6439 and https://github.com/nodejs/node/pull/6375

My proposal is to use the new `process.emitWarning` API to treat unhandled rejections as warnings (but not errors), this is effectively a relaxation of the "log on unhandled rejection" option since it is treated as a warning and not an error and can be handled with either installing a `unhandledRejection` handler _or_ by disabling warnings in Node (for example in production). It is meant to address the ergonomics point of view.

@Fishrock123 's proposal is to abort on unhandled rejections once we are absolutely sure they are unhandled (by hooking into garbage collection), this is effectively a relaxation of "throwing an error" option addressing the problem in that option of not being sure 100% it is thrown - aligning it with regular exceptions.

Please drop by at these issues and leave a comment.

**Comment by chrisdickinson** [comment]
One more proposal to add into the mix: crash by default on unhandled rejection; add a `--allow-unhandled-rejection` flag to allow Node to continue past unhandled rejections.

**Comment by benjamingr** [comment]
Just pointing out that that proposal is not exclusive with either existing proposals - it generalizes the "throw on GC" proposal and replaces the "warn on unhandled rejection" proposal by replacing warnings with exceptions which is semver-major anyway (and probably with a safety version) :)

**Comment by yosuke-furukawa** [comment]
IMHO, I would like to detect that I forget to handle `unhandledRejection` carelessly, so I'm +1 to @benjamingr plan. But I have a little curious, that behavior is different `uncaughtException` .

For example,

```js
const fs = require('fs');
fs.readFile('foo/bar/buz', (err, data) => {
  if (err) {
    throw err; // this err will occur uncaughtException, so process will exit
  }
});
```

According to this PR, the above example will change after merged.
https://github.com/nodejs/node/pull/5020

```js
const {readFile} = require('fs').promise
readFile('foo/bar/buz').then((data) => {
}).catch((e) => {
  throw e; // this err will occur unhandledRejection, but process won't exit and log error.
});
```

**Comment by benjamingr** [comment]
@yosuke-furukawa both @Fishrock123's idea (log on GC) and @chrisdickinson's idea (log on unhandled rejection synchronously) address that part and neither are exclusive to this PR (or each other).

Note that your callback example isn't really what code looks like- callbacks would swallow exceptions in this case (duh) it just looks different: 

```
const fs = require('fs');
function readBuzz(cb) { 
  fs.readFile('foo/bar/buz', (err, data) => {
    if (err) {
      if(cb) cb(err); // operational exceptions are propagated through err param
    }
});

readBuzz((err, data) => {
   // I don't have to pass a callback at all, and if I do I don't have to check the data parameter 
})
```

This PR is mostly for programmer errors to not go silenced - the path we go afterwards (throwing on unhandled rejection synchronously, on unhnalded rejection or not at all) is important - but orthogonal to this specific PR. 

I want to solve the swallowed errors by default problem for 99% of promise users first - I believe this will provide very useful data on what users expect and based on the reception of this PR we'll move forward with fixes that address the theoretical side of the problem (if at all) in the future.

**Comment by sholladay** [comment]
As a library author looking upstream, anything but crashing is very frustrating. I really wish promise errors were modeled after `uncaughtException`.

This program works perfectly, as in it does exactly what I expect. It crashes and burns nice and loud. I made a mistake and not only does Node call me out on it, but it also refuses to let me continue being an idiot.

```js
'use strict';
setTimeout(() => {
    setTimeout(() => {
        console.error('Oh dear. Everything is screwed.');
    }, 10);
    const x = {};
    x.oopsie();
}, 100);
```

This program ought to do the exact same thing, but it just keeps going no matter how many things go wrong.

```js
'use strict';
new Promise((resolve) => {
    setTimeout(() => {
        console.error('Oh dear. Everything is screwed.');
    }, 10);
    const x = {};
    x.oopsie();
    resolve();
});
```

How many of those bugs are lurking on npm? Think of how thankful people will be if you help them fix their programs.

What if `uncaughtException` had started off the way `unhandledRejection` is now? People wouldn't be able to get anything done in Node. Would it never get fixed because people had at least attempted to build things on top of it despite its flaws?

This is why we have semver and major versions. It's why Node Forward became a thing. And it's why many of us left PHP behind. That is, _so we can make progress_.

How much of not doing this is FUD vs real-world experience and facts? How often do people add late `.catch()` handlers vs commit serious programming mistakes? In particular, what would the experience be like for a developer who used a build of Node that throws rejections for a week?

I will try to keep an open mind, but I believe it would overwhelmingly lead to bug fixes and kudos for having a more trustworthy system.

**Comment by isiahmeadows** [comment]
Just a question: couldn't this be implemented as a `v8::Persistent` finalizer on the promise in C++ land, just constructing a string from the error during finalization and then returning back to JS land to call `console.error`?

This avoids the issue of a promise being used a day later after an error was thrown and previously unhandled until then (a live, strong reference technically still existed).

**Comment by isiahmeadows** [comment]
@sholladay I'm against crashing unless the promise was actually unhandled when GC hits it. Something is better than nothing, but crashing over that is not preferable, especially if it's in a GUI program (e.g. in NW.js or Electron) where a handler was added on a tick after the promise was resolved. Warning is okay in my opinion, since that's not very quickly user-visible, but crashing altogether would be incredibly terrible for UX.

**Comment by ChALkeR** [comment]
https://github.com/nodejs/node/pull/8217 and https://github.com/nodejs/node/pull/8223 landed.

**Comment by sholladay** [comment]
@isiahmeadows this is about the _default_ behavior. That makes a world of difference.

The default handler for `unhandledRejection` shouldn't let the program fumble on past programming errors just because 0.00001% of projects (fact check?) want late `.catch()` handlers on a future tick. That's crazy.

The happy path should protect against programming errors and be symmetrical with `uncaughtException`.

I fully support an opt-out. Today, if you decide to be reckless you can do this.

```js
process.on('uncaughtException', (err) => {
    // Ignore the error, because we don't care.
});
```

That's how this should work, too. One very nice side effect about this approach is that the developer's intentions are clearer. Just as I actively look for and avoid code that listens for `uncaughtException` unless it's _very_ high quality, I would do the same for `unhandledRejection`.

**Comment by ljharb** [comment]
@sholladay The default path should be compliant with the language spec, which specifically allows the use case of later-added catch handlers.

**Comment by benjamingr** [comment]
@ljharb NodeJS does and will continue to allow late catch handlers. 

The agreed on policy is:

 - Abort when an unhandled rejection can never be handled (a GC hook @Fishrock123  is working on). 
 - Warn on late catch handlers since they're _likely_ a mistake. This is easily opt-out for users with legitimate use cases for late bindings but should really help the other 99%. 

This should please everyone involved since all use cases are supported and we don't introduce abortive semantics unless we're 100% sure the promise is unhandled. Libraries can even decide to opt out of the default behavior.

**Comment by billinghamj** [comment]
It might be worth having something like:

```js
process.on('promisePossiblyUnhandled', error => {
});
```

So it isn't the same as an unhandled exception where the process will exit, and it would still be fired in the case of a late handler.

**Comment by billinghamj** [comment]
How would you opt-out of the late handler warning though? Would you need to have an additional method on the promise indicating that you are intending on handling it late?

**Comment by benjamingr** [comment]
@billinghamj not only would it make sense it's actually the case right now. An `unhandledRejection` event is fired and a `rejectionHandled` if the rejection is handled through a late catch.

You can opt out of the late handler warning in two ways: either you disable the warnings or you add your own `unhandledRejection` handler:

```js
process.on('unhandledRejection', (err) => {
    // Ignore the rejection, because we don't care.
});
```

**Comment by sholladay** [comment]
Has there been any discussion about having a `--throw-rejection` like how we have `--throw-deprecation`?

I currently do this in every program:

```js
process.on('unhandledRejection', (err) => {
    throw err;
});
```

I find it so necessary that I [published it](https://www.npmjs.com/package/throw-rejects). And I don't think a warning is going to pacify me. But if I at least had a bash alias that I could use instead of vanilla node, I would definitely spend less time debugging.

**Comment by benjamingr** [comment]
There has been discussion but this will likely be solved when we start throwing on unhandled rejections based on garbage collection. 

Basically the behavior you describe will be automatic for rejected promises that are GCd.

**Comment by Jeff-Lewis** [comment]
@sholladay Allowing `unhandledRejections` to continue and not kill the process feels OK if you think of Promises as also being a `try {} catch {}`. From one perspective, those that learn how to use Promises from [Bluebird](https://github.com/petkaantonov/bluebird/) which [describes them that way](http://bluebirdjs.com/docs/why-promises.html), it feels natural.

In this case, an `unhandledRejection` is just like a `try {}` without a `catch {}`. Now is that good programming, maybe not, but `try {}`s without `catches{}`s are definitely intentionally used.

Also, here's one way of keeping track of `unhandledRejections`.
``` javascript
let unhandledRejections = new Map();

/**
 * Promises/Bluebird specific global event.
 * NOTE: event name is camelCase as per node convention
 */
process.on('unhandledRejection', function(reason, promise) {
	unhandledRejections.set(promise, reason);
	console.log('Unhandled Rejection. Total:' + unhandledRejections.size, reason, reason.stack, promise);
});

/**
 * Promises/Bluebird specific global event.
 * NOTE: event name is camelCase as per node convention
 */
process.on('rejectionHandled', function(promise) {
	unhandledRejections.delete(promise);
	console.log('Rejection Handled. Total:' + unhandledRejections.size, promise);
});
```

**Comment by sholladay** [comment]
> an unhandledRejection is just like a try {} without a catch {}

When you say _without_, I assume you mean with an empty `catch`. And yes, this is precisely what's going on. Synchronous code behaves correctly: you have to opt-in to ignoring errors with empty catch blocks. As it stands, promises are implicitly wrapping everything in a `try` with an empty `catch`. Imagine if I proposed doing the same thing to synchronous code.

Anyway, I appreciate that everyone is at least listening. And there seems to be movement in the right general direction.

**Comment by jrop** [comment]
Exactly what @sholladay said.  I am not sure why there is so many differing opinions regarding this topic, as the current behavior of node/v8 of swallowing the errors is a terrible default behavior, though one that I know the developers are working hard to address.

Currently the team I am on utilizes promises heavily (using generators + co), and every time we create an asynchronous function, it looks something like this:

```js
const myFunction = co.wrap(function *() {
    try {
        // ... async code here ...
    } catch {
        console.error(e.stack)
    }
})
```

NOTE: many of the async functions we create are not called from other async functions, so we do not have the option of letting the error bubble-up to a "top-level" handler.

That is a lot of boilerplate for each function to implement just so that my code does not swallow errors.  I truly cannot wait until the default behavior of rejected promises die loudly (upon GC, as noted).

**Comment by ljharb** [comment]
@jrop it's language-mandated behavior. Whether you like it or not doesn't affect that to be called a JS engine, it has to implement JS.

**Comment by jrop** [comment]
@ljharb Do I understand you correctly in that you are saying that the ECMAScript specification dictates that unhandled rejections are to be swallowed?

**Comment by bmeck** [comment]
@ljharb the abrupt completion handling is at [PromiseReactionJob](https://tc39.github.io/ecma262/#sec-promisereactionjob) and [RunJobs](https://tc39.github.io/ecma262/#sec-runjobs). When using [HostReportErrors](https://tc39.github.io/ecma262/#sec-host-report-errors) the operation *must* complete normally (assuming it can complete). If we `abort` on GC I don't see a violation unless I am mistaken.

**Comment by ljharb** [comment]
@bmeck oh totally, abort on GC is totally fine - I'm saying that's the only place you should abort.

**Comment by jrop** [comment]
@ljharb it sounds like we are agreeing.  I am also in favor of aborting on GC.

**Comment by benjamingr** [comment]
I was under the impression we've reached consensus already with https://github.com/nodejs/promises/issues/26#issuecomment-246172860 - please let me know if anyone disagrees I was under the impression we're in agreement :)

**Comment by billinghamj** [comment]
@benjamingr Re: the warnings. How would you "disable the warnings" as you mentioned above?

Additionally this is fine if you're writing an application:

```js
process.on('unhandledRejection', (err) => {
    // Ignore the rejection, because we don't care.
});
```

But what if you're writing a library, which is legitimately catching late? You don't have control over the runtime, and you shouldn't override any process events, but you don't want warnings to appear on the console.

**Comment by ljharb** [comment]
@billinghamj `var promise = maybeRejected(); promise.catch(() => {}); return promise;`?

**Comment by billinghamj** [comment]
Ah yes, makes sense. Just catch it without any action at first, then run `.catch` again later.

**Comment by otakustay** [comment]
That's not a good solution, for library developers, catch a promise with empty function could lead errors to be swollowed, furthur `.catch` call could never receive the error

The "Early throw, late catch" principle tells us not to catch any error if we cannot correctly handle it, so library developers usually SHOULD NOT handle errors (and an empty catch block is the worst case)

**Comment by benjamingr** [comment]
@otakustay I'm not sure you're talking about the same thing @ljharb and @billinghamj  are talking about.

What @billinghamj asked was how to signal (as a library) that an error might be thrown late. What adding a catch handler to the promise while forking it and returning _the original_ does is mark the error as not being an unhandled rejection.

No one is suggesting library authors actually suppress errors - just mark `catch`es where it is legitimate to catch the error very late or even not at all. The error can still be handled.

**Comment by otakustay** [comment]
Sorry I'm really mistaking the purpose and forgetting the **fork** part of promise

**Comment by shenanigans** [comment]
So *point A* like most of the people in the conversation I'm only here because I suddenly have to suppress a default and I don't like that. This is a very disruptive change and I'm dealing with it in my current project by taking the straightest path from here to having control of my console output again. Jumping straight to `exit (1)` on this update would've been frankly less aggrevating.

*Point B* Can we really not hold an error through the current tick? It's currently impossible to catch an error thrown synchronously by a Promise because it occurs before the constructor closes. Let's get one thing straight right now: This is not a "late catch" scenario. It's the earliest possible catch scenario and it's still too late. Quite the embarrassing API mistake in my opinion.

*Point C* My current project accepts an optional callback but always returns a Promise. I should at least be able to indicate that I don't care about unhandled rejections if nobody ever listens to the Promise at all.

*Point D* Am I really the only person that thinks ES6 is garbage?

**Comment by ljharb** [comment]
@shenanigans re point B, it should only be erroring when the promise gets garbage collected - in other words, when you have no possible way to handle it later. re point C, `promise.catch(); return promise;`. re point D, there's no need for negativity - there's nothing garbage about it, and that kind of acrimony doesn't belong on this thread, or this site.

**Comment by shenanigans** [comment]
@ljharb I'm only working with Promises at all (see point D) because I'm writing a drop-in replacement and apparently async functions returning `undefined` isn't hip anymore. With that being my level of enthusiasm for this feature, the absolute last thing I wanted to deal with was trash in `stdout` from a breaking change in the engine. That's not enough though, the only way to suppress it is to suppress it globally for anyone who includes my lib. You bet I'm feeling negative and acrimonious about that.

I'm also being overly general, though. Symbols are pretty cool.

**Comment by ljharb** [comment]
@shenanigans as I [explained](https://github.com/nodejs/promises/issues/26#issuecomment-257487824) re point C, there is an *incredibly trivial* way to suppress it for a single promise: call `.catch` on it, and then return the promise (not the one returned by `.catch`, the original one).

**Comment by benjamingr** [comment]
> So point A like most of the people in the conversation I'm only here because I suddenly have to suppress a default and I don't like that. 

I have actually surveyed about 50 developers anonymously last week about the feature and default and I received overwhelmingly positive response (45 out of 50 being in favor, 4 not feeling strongly and 1 not liking it) about changing the default from swallowing errors. I maintain that not swallowing errors is a much preferable default to what we did before.

> Point B Can we really not hold an error through the current tick? It's currently impossible to catch an error thrown synchronously by a Promise because it occurs before the constructor closes. 

Both the current implementation and the GC based one allow you to attach catch handlers inside things like `nextTick` - you can catch it synchronously but it will also work if you catch it asynchronously but before more IO happened.

> Point C My current project accepts an optional callback but always returns a Promise. I should at least be able to indicate that I don't care about unhandled rejections if nobody ever listens to the Promise at all.

As @ljharb said - it's very easy to suppress this by adding a `promise.catch(() => {}); return promise`. However, if you use a library like bluebird you can also opt out of these warnings for your library but not in general. I just want to point out that a lot of people find these warnings very helpful and I have personally gotten overwhelming positive feedback regarding this change from the community.

> Point D Am I really the only person that thinks ES6 is garbage?

I'd rather not reply directly to this as it's not directly related to the conversation. You are welcome to voice your concerns to the technical committee at https://esdiscuss.org or by signing to the mailing list https://mail.mozilla.org/listinfo/es-discuss .

**Comment by rvagg** [comment]
My non-scientific reading of reactions we're getting to the change is mostly positive too. But of course as with any change we have people who are upset about it too. I understand that changing default behaviour will always be upsetting for some people depending on your personal preferences but on balance this change is wholly in line with standard expectations of error handling within Node.js, expectations that have been set by unhandled exception and unhandled `'error'` behaviour. It would also be entirely consistent if we found a way to do the same with unchecked `err` parameters on callbacks.

**Comment by leodutra** [comment]
Great. If we gonna distort defaults for better quality of life... I'll
strike once more: single threaded JS is old browser bullshit and makes JS a
less competitive language.

All async processment should run in a libuv thread. It's almost absurd we
don't even try to see the benefits... Parallel is better than concurrent.

**Comment by leodutra** [comment]
This would make Node and JS in Node context way more powerful than Nashorn
and other pure engines and put it side by side with Akka and OTP usage.

Freeing the Event Loop from this responsibility is almost a natural choice.

Maybe we'll have to rethink errors (and I don't really see a hard change on
this). But at least, I think community should consider exploring more the
libuv capabilities or we will have a halved  awesome potential forever.

Promises should be streamed and async await should hide a streamed
resolution. This is not my idea... C# and others do it.

Please, please, please think in it.

**Comment by isiahmeadows** [comment]
I invite you also to check out the es-discuss mailing list. It's the best
place to talk about those kinds of things. Speaking of which, parallelism
itself has come up recently (past week or so). So do check it out.

https://esdiscuss.org/
https://mail.mozilla.org/listinfo/es-discuss

**Comment by benjamingr** [comment]
> Great. If we gonna distort defaults for better quality of life... I'll strike once more: single threaded JS is old browser bullshit and makes JS a less competitive language.

And again, there are efforts to parallelize JS where this sort of input might be appropriate although to be completely fair I would probably try to approach those bodies with a less argumentative tone. This is about bettering the life quality of developers - let's stay positive. I'll repeat my recommendation for esdiscuss.

> All async processment should run in a libuv thread. It's almost absurd we don't even try to see the benefits... Parallel is better than concurrent.

Async processing does run in a libuv threadpool and notifications are passed to node through event listeners. One of the more common misconceptions is that node is single threaded - that is certainly not the case - node just has a single JS thread. In addition, the `cluster` module lets you create several JS processes and communicate between those.

> This would make Node and JS in Node context way more powerful than Nashorn and other pure engines and put it side by side with Akka and OTP usage.

While I appreciate both OTP and Akka I find them better than node for some things and a lot worse for others. They use a paradigms and put the emphasis in different places. Node is, like most software a result of many tradeoffs. For a lot of what I do it's fantastic, for a lot I wouldn't consider it.

> Maybe we'll have to rethink errors (and I don't really see a hard change on this). But at least, I think community should consider exploring more the libuv capabilities or we will have a halved  awesome potential forever.

Consider looking at the `libuv.js` project which exposes a much lower level API. I'm not sure what's the current status is but the great thing is that everything is done in the open and you're welcome to explore possibilities of changing how node works - I'll gladly take.

As for errors - I definitely think we need to rethink errors. Easier said than done :)

> Promises should be streamed and async await should hide a streamed resolution. This is not my idea... C# and others do it.

I'm not sure what that means but I'm very intimate with how async/await is implemented in C# and it's pretty similar (sans things like synchronization context, and C#.next efforts to replace GetAwaiter with something nicer). 

There are two proposals for streaming at the moment - async iterators and observables. Just to repeat, this is not the place to discuss these things - esdiscuss and GH issues on https://github.com/tc39/proposal-observable and https://github.com/tc39/proposal-async-iteration are. 

It sounds like you sound passionate about this - by all means there are a lot of ways to contribute to Node, JavaScript and the ecosystem. Should you ever find yourself "stuck" ping me - there is a lot more work to go around than people able to spend time and do it.

-----

On a less positive note - whenever you comment here it pings 22 people via email. So this is my last reply here that's not directly about the default.

**Comment by bminer** [comment]
@benjamingr - Does anyone have an update on this?  I read in https://github.com/nodejs/node/pull/12734 that the *new* decision is to exit Node when an unhandled Promise rejection occurs **and** there is no `unhandledRejection` handler.

I agree with this behavior (long debate on https://github.com/nodejs/node/issues/830), and I think that crashing when a rejected Promise is garbage collected is... well... sloppy... especially when you're just trying to make <1% of Promise users (those employing "late" catch handlers) happy.  It introduces the *possibility* of false negatives, and it may cause programs to fail more often when under memory pressure.

As mentioned numerous times, users could simply:

```javascript
process.on("unhandledRejection", () => {});
```

to override default behavior and add catch handlers late.  Just like you can do this if you want to live on the edge:

```javascript
process.on("uncaughtException", () => {});
```

IMHO, I think it's very important for asynchronous and synchronous *uncaught* errors to have the same *default* behavior.  Currently, I just do this at the top of my `index.js` file:

```javascript
process.on("unhandledRejection", (reason) => {throw reason;});
```

I don't care if I need to continue that practice or not, but I certainly have my opinions formulated about unhandled Promise rejections.  What's a tad frustrating for me as a Node user is that this debate has been going on for 3+ years.  :(  I feel like we have all expressed our opinions, and all of the facts are out there.

That said... what is the final consensus?  Any idea when the decision will impact Node 8 or Node 9 users?

**Comment by flying-sheep** [comment]
fail early is a good and consistent concept. i don’t see why async stuff should be treated as some kind of low priority background stuff that can be allowed to fail.

**Comment by ljharb** [comment]
Because the language explicitly is designed to, and does, allow late failures (https://github.com/nodejs/promises/issues/26#issuecomment-188879863)

**Comment by bminer** [comment]
@ljharb - We've each argued our sides.  I can see the use case for late catch handlers, so I understand your frustration.  Still, please keep in mind that we are only discussing **default** behavior; no one is saying you can't have late catch bindings.  After all, it's [one line of code to prevent the default behavior](https://github.com/nodejs/promises/issues/26#issuecomment-349021518), and more complex logic could be split into a 3rd party module.  I'd be happy to argue further, but let's take it offline -- [just email me](https://github.com/bminer).

At this point, I'd just like someone to make a decision and stick with it.  :)  It's pretty clear what I'd prefer, but I'll accept whatever is best for the Node community as a whole.

**Comment by jrop** [comment]
@bminer 
**Every** catch handler has the potential to be a late error handler ([the executor runs synchronously before the Promise constructor returns](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)):

```js
let state = 'pending'
const promise = new Promise((resolve, reject) => {
  // this executor runs before anything else,
  // and since it is synchronous, the promise-state
  // is rejected before we can attach the handler to it below:
  state = 'rejected'
  reject(new Error('uh-oh'))
})
console.log(state) // 'rejected'
promise.catch(e => { // technically attached "late"
  console.error('my handler:', e)
})
```

This is because Promises are not true monads: instead of "lazy-loading" them (when a handler is attached via `then`/`catch`), promises are optimistically executed.

This is why I like the GC approach.

**Comment by benjamingr** [comment]
@jrop that's not technically attached late in Node's regard, attached late means I/O already happened:

```js
var p = Promise.reject(new Error());
p.catch(() => {}); // not late
for(var i = 0; i < 1e9; i++); // busy loop
p.catch(() => {}); // still not late
process.nextTick(() => {
  p.catch(() => {}); // still not late
});
Promise.resolve().then(() => {
  return new Promise(resolve => process.nextTick(resolve));
}).then(() => {
  p.catch(() => {}); // still not late
});
fs.readFile("foo.txt", (err, data) => {
  // this is late
});
```

**Comment by benjamingr** [comment]
Also:

> This is because Promises are not true monads: instead of "lazy-loading" them (when a handler is attached via then/catch), promises are optimistically executed.

Promises are not "executed", promises are the results of an already existing action. Monads have nothing to do with laziness (haskell thunk laziness does, but that's orthogonal). Monads are just things that implement the monad interface (belong to the monad typeclass). The requirements are having an unary return (Promise.resolve), a type constructor (Promise) and a bind operator (Promise.then) with the signature `M a -> (a -> M b) -> M b` (`then` with the first `M` being `this` and the single function overload - `Promise<a>::then(a -> Promise<b>) -> Promise<B>`.

**Comment by dead-claudia** [comment]
@jrop And to add to @benjamingr's comment, what makes Promises not monadic is that you can't nest resolved ones. Specifically, the monadic laws require that `Promise<Promise<T>>` must be possible (and by proxy, `Promise.resolve(value).then(x => Promise.resolve(x))` be equivalent to `Promise.resolve(Promise.resolve(value))`, but not `Promise.resolve(value)`), but JS doesn't allow it. Has nothing to do with execution, but the types themselves. If JS didn't auto-resolve thenables, then it'd be a little more conforming. It was a point of conflict during the Promises/A+ design discussions, but not really anymore beyond various purists (and for them, alternatives exist).

**Comment by spion** [comment]
Actually the laws don't hold even if you forbid nesting

```
Promise.resolve(a).then(f) <equivalent to> f(a)
```

Immediately this is not true, since `f` can throw or return a non-promise.

But lets say we restrict the type of `f` so that it can only return promises. A more fundamental problem that cannot be solved without breaking the A+ spec still remains. The executed side effect is deliberately not the same. The left side will execute with one tick delay on the microtask queue, while on the right side the side effect executes immediately. So the side effects are not equivalent. 

This breaks application of the law in [equational reasoning][1]. If we write `f(a); f(b);` and then substitute based on the law with `Promise.resolve(a).then(f); f(b);`, in the first case the side effects will start in order; in the second case, they will start in the reverse order. 

Equational reasoning is the main reason that the laws exist, so I believe the above is a sufficient reason to say that the law doesn't hold, and therefore a Promise (even the variant that would allow nesting) is not a monad. In fact, its very difficult to uphold any sort of equational reasoning if there are any eagerly executed side effects present, since the order of execution is defined by the order of evaluation!

[1]: http://www.haskellforall.com/2013/12/equational-reasoning.html

**Comment by leodutra** [comment]
I vote for another async hook alternative, similar to `process.on` + existing error upstream propagation.

Considering async/await, the stream is declared async and run in a async way. Their starter should be able to try {} catch(err){} if necessary (as it is). And the process, as a "container" of operations, should be notified of it's internal events (of any kind).

Beautiful? Maybe not for some practical eyes... but essentially aligned with general concept of processes, events and "green threads".

I still think Node.js should have a thread runner lib, using Promises/events for communication (like service workers or common Java/Go threads)... but this is another discussion. Currently, It's weird to think in a Node.js Actors model without it. And if there's someone trying to implement this way, I don't really wanna see. :nauseated_face: 

+: I think logging unhandled on stderr is very valid... making possible to flag for no logging.
